### PR TITLE
Google component filters

### LINKED ISSUE #426: Google
In my project I have extended the GoogleMapsProvider to have a setter for component filters.

https://developers.google.com/maps/documentation/geocoding/#ComponentFiltering

If I'd create a PR, would you such kind of additional functionality?

### PR DESCRIPTION
**Pull request by mvcaaa**
Fully based on pull/427

Tests added
Some fixes to original code

Closes #426

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,7 @@
+.idea/
```
Comment (Baachi) [inline .gitignore]: This should go to your global gitignore.

File: .gitignore [hunk 2]
```diff
@@ -1,5 +1,7 @@
...
 vendor/
 composer.lock
 composer.phar
 phpunit.xml
 php-cs-fixer.phar
+tests/.cached_responses/
```
Comment (Baachi) [inline .gitignore]: This should be reverted, too

**Comment by willdurand** [comment]
@geocoder-php/geocoder WDYT?

**Comment by dereuromark** [comment]
Useful addition! :+1:

**Comment by toin0u** [comment]
It's a good addition - the component filters improve really the result :) :+1:

**Comment by willdurand** [comment]
@mvcaaa could you rebase this PR please?

**Comment by mvcaaa** [comment]
@willdurand Done

**Comment by willdurand** [comment]
@mvcaaa sorry to bother you one more time, but could you:

1. remove this commit https://github.com/geocoder-php/Geocoder/pull/466/commits/50dfe8196a360eaf683d394c25a2caff18db7b3e
2. squash your commits to a single one with a new commit message

**Comment by mvcaaa** [comment]
Any feedback ?

**Comment by Nyholm** [comment]
I know it has been a while since this PR was submitted. But I can't find the documentation of component filtering. The link you provided does not work... (or Im blind)
https://developers.google.com/maps/documentation/geocoding/#ComponentFiltering

**Comment by LordBaine2000** [comment]
They changed it at some point to https://developers.google.com/maps/documentation/geocoding/start#ComponentFiltering

File: src/Geocoder/Provider/GoogleMaps.php [hunk 3]
```diff
@@ -167,11 +192,21 @@ private function buildQuery($query)
...
+                return $key.':'.urlencode($componentFilters[$key]);
+            }, array_keys($componentFilters))));
+        }
+
         return $query;
     }
 
     /**
      * @param string $query
+     * @return \Geocoder\Model\AddressCollection
```
Comment (Nyholm) [inline src/Geocoder/Provider/GoogleMaps.php]: Remove this. It is not a part of the PR and it is no longer valid.

File: tests/Geocoder/Tests/Provider/GoogleMapsTest.php [hunk 4]
```diff
@@ -427,4 +427,64 @@ public function testGeocodeWithInvalidClientIdAndKeyNoSsl()
...
+    public function testGeocodeWithComponentFiltersInvalidZipCode()
+    {
+        $provider = new GoogleMaps($this->getAdapter(), 'fr-FR', 'Île-de-France');
+        $arrFilters = ['country' => 'fr', 'postal_code' => '00000'];
+        $provider->setComponentFilters($arrFilters);
+
+        $provider->geocode('10 avenue Gambetta, Paris, France');
+    }
+
+    public function testGeocodeWithValidComponentFilters()
```
Comment (Nyholm) [inline tests/Geocoder/Tests/Provider/GoogleMapsTest.php]: Would this test be different without setComponentFilters?

I suggest that you only write a **unit** test that make sure the correct parameters is added on the URL. We already have functional tests that make sure the response handing is proper.

File: src/Geocoder/Provider/GoogleMaps.php [hunk 5]
```diff
@@ -141,6 +151,21 @@ public function setRegion($region)
     }
 
     /**
+     * Add components for filtering.
+     * https://developers.google.com/maps/documentation/geocoding/#ComponentFiltering
```
Comment (Nyholm) [inline src/Geocoder/Provider/GoogleMaps.php]: Really good that you provide the link, but it should be updated.

File: src/Geocoder/Provider/GoogleMaps.php [hunk 6]
```diff
@@ -133,6 +138,11 @@ public function getName()
         return 'google_maps';
     }
 
+    /**
+     * @param $region
```
Comment (Nyholm) [inline src/Geocoder/Provider/GoogleMaps.php]: This doc block is not part of the PR. You can keep this but try to avoid "code cleanups" in the same PR as features.

File: tests/Geocoder/Tests/Provider/GoogleMapsTest.php (hunk 4 above)
Comment (mvcaaa) [inline tests/Geocoder/Tests/Provider/GoogleMapsTest.php]: ComponentFilters are 100% optional - so without it this test will not  be different, but it will fail 100%

Not sure how i can redo this - any suggestions are welcome.

Comment (Nyholm) [inline tests/Geocoder/Tests/Provider/GoogleMapsTest.php]: Instead of doing this functional tests we could do unit tests that verifies that the query URL has changed when using setComponentFilters. We do not actually need to execute the query and verify the result.

Comment (mvcaaa) [inline tests/Geocoder/Tests/Provider/GoogleMapsTest.php]: Yes, you are 100% right, but
Here is my concern: query is formed by private function buildQuery(and only there). I can test it, but it isn the perfect solution, kind of, i think

Comment (Nyholm) [inline tests/Geocoder/Tests/Provider/GoogleMapsTest.php]: What if you mock the HttpClient and verify that the sendRequest function has a request with the preferred query?

Comment (mvcaaa) [inline tests/Geocoder/Tests/Provider/GoogleMapsTest.php]: It can be done, sure. Will do if no one was willing to disagree or suggest something better.
I think, buildQuery must be refactored also - but i will not do that. It is way outside my small tiny PR.

Comment (Nyholm) [inline tests/Geocoder/Tests/Provider/GoogleMapsTest.php]: > I think, buildQuery must be refactored also - but i will not do that. It is way outside my small tiny PR.

Yes it is outside the scope. 

I do not think you have to refactor anything. Just update the test code.

Comment (arubacao) [inline tests/Geocoder/Tests/Provider/GoogleMapsTest.php]: @mvcaaa i did something silimar for googles `result_type` & `location_type` and wrote a unittest that checks the url.
Check out here: https://github.com/arubacao/Geocoder/blob/8f322d551b486c8c3bf6fd7fe84caaec6364797e/tests/Geocoder/Tests/Provider/GoogleMapsTest.php#L304

**Comment by Nyholm** [comment]
Ping, will you complete the work on this PR or should it be closed?

**Comment by Nyholm** [comment]
Im closing because inactivity. Feel free to create a new PR. I like this feature.
